### ISSUE TITLE
Prevent sending private toot if no-one is mentioned.

### ISSUE THREAD

**Issue by roipoussiere** [tootsuite/mastodon#7073]
It's a common mistake for me: sometimes I send a toot supposed to be public, then I realize later that my toot is a private message and no-one is mentioned (so only me can read it). 🤦‍

I note that on the Mastodon UI, the privacy-level switch go back to *public* after a DM is sent, so it avoids this behavior a little bit, but it's not the case on other Mastodon client.

Is it possible to prevent sending private toot if no-one is mentioned, at the *server* side, in this way any Mastodon client can display an error message?

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by BaptisteGelez** [comment]
Actually, it can be useful if you want to test things without spamming everyone.

**Comment by MightyPork** [comment]
this it's commonly used for testing toots before sending as public.

also, for silent rants. This would be a regression

**Comment by Gargron** [comment]
On the other hand it *can* be used to sneakily spam the server. You could create two accounts and spam between them even if made DMs-to-nobody impossible, but it would involve a bit more effort. I'd say I'm in favour of making DMs-to-nobody fail?

**Comment by MightyPork** [comment]
@Gargron do you have any evidence of this happening, compared to people actually using it for legitimate purposes like testing?

**Comment by Gargron** [comment]
No, but the theoretical possibility is worrying. I also don't have evidence of many people using it for testing, tbh.

**Comment by trwnh** [comment]
Wouldn't any evidence of such be... yknow... private? ;)

I use direct-to-nobody occasionally to post reminders to myself, e.g. to follow up on something I posted but without spamming my profile.

**Comment by Technowix** [comment]
I'm actually using it to test many things personally.
Also, if the spam is the issue, rate limiting is actually the solution...

**Comment by TheKinrar** [comment]
Creating two accounts is really easy, I don't think removing DMs-to-nobody will change anything to the possibility of spamming.

**Comment by wryk** [comment]
I use DMs-to-nobody for testing and silent rant, don't remove them.

Because someone could abuse a feature to spam are you going to remove :
* followers-only without followers
* followers-only/DM with an alt account

I don't think so.

**Comment by Hasimir** [comment]
> On the other hand it can be used to sneakily spam the server.

For the purpose of what?  DoS?  If that's actually a vector of attack of concern then you can either:

 1. Enforce rate limiting.
 2. Suspend or block the user.

> You could create two accounts and spam between them even if made DMs-to-nobody impossible, but it would involve a bit more effort. I'd say I'm in favour of making DMs-to-nobody fail?

Pointless and it's good to have a default of not posting an incorrectly addressed DM just returning to self.

As for the self notification via DM thing, which could be very useful (like setting an automated reminder to take your damned vitamins ... which reminds me, where are mine ...) ; there's always the option of explicitly setting the DM to your local account on that instance,  That works too and should definitely remain, even if you fail the absence of any target accounts.

**Comment by Dhveszak** [comment]
Just create a warning -like the box-thing that says "your account isn't private" when you switch to followers-only privacy- when you don't tag anyone in a DM.
I use DMs-to-nobody for testing and notes to myself, but also to save drafts for later, they're really useful.

**Comment by Sylvhem** [comment]
I usually use DMs-to-nobody to test how a toot is going to be formated before sending it. It comes in handy a lot when the newest meme is something graphical (like the sign bunny or the one with the starfield).
Also, I sometimes use them as a way to make draft before a complicated series of toots.

**Comment by roipoussiere** [comment]
Ok, thank you for the clarification I didn't know that DM-to-myself was actually a feature use by some people. :D So this is more about how Mastodon clients handle this behavior.

I created [an issue on the Twidere repo](https://github.com/TwidereProject/Twidere-Android/issues/1108) (the client I use).

An alternative option for the Mastodon UI can be a tooltip on the *Send toot* button saying: "Hey your are going to send a message visible only by yourself".

**Comment by SoniEx2** [comment]
see #4877 and #4885

**Comment by NekomataEsp** [comment]
I also use dms to myself as a way to test toots that rely on visual things

**Comment by codl** [comment]
Yeah what everyone says, I use it for checking formatting 

Maybe there ought to be a feature to preview a toot before sending?

**Comment by nightpool** [comment]
I use DMs to myself to remind me to do things later/in the morning

**Comment by brickcamp** [comment]
Could there be at least a little visual hint indicating that there are no mentions in a DM? 

I replied to a DM weeks ago and accidentally deleted the mention of the sender while drafting my response. The message never arrived (of course) but I noticed this only today a few weeks after. Maybe it could be indicated in the hint text above the editor? Like ...

![Bildschirmfoto-20190413145117-461x340](https://user-images.githubusercontent.com/49322756/56079976-b072d180-5dfb-11e9-82dc-041d346e81cf.png)

..., maybe change it to another background colour if it's the case? Or as a second separate text box or whatever looks best. Think it wouldn't hurt but help some first-time users (and dumb ones like me :laughing: )
